### ISSUE TITLE
Add metrics to ingest pipeline

### ISSUE THREAD

**Issue by cawarren** [Recidiviz/recidiviz-data#59]
From a comment Josh made on PR #46 -

- Number of new inmates, records, and snapshots created per session
- Number of existing inmates, records, and snapshots updated per session
- Number of each field that was updated in a new snapshot per session (would be super handy to help understand the latent volatility in a region's source data)
- Infrastructure metrics related to the task queue itself, e.g. number of retries, average idle time per worker, average wall time per task, etc.

First step will be to add metrics similar to the above to a session (there can be many sessions to a full scrape). Follow-up step will be aggregating up to a scrape.

**Comment by cawarren** [comment]
_This is part of ingest generalization work_

Scope update:
- Read and determine how to use Google Stackdriver for custom metrics in Google App Engine. Some relevant links:
  - [Using Stackdriver logging in App Engine apps](https://cloud.google.com/appengine/articles/logging)
  - [Stackdriver : Using Custom Metrics](https://cloud.google.com/monitoring/custom-metrics/)
- Enable Google Stackdriver support in Cloud Console, set up configuration
- Add `metrics` module under `ingest` package (see the design doc's [Modules section](https://docs.google.com/document/d/11fv-hLwLOFcM5XSE16lYne5hnqFzcQK1vwvTBVVyhn8/edit#heading=h.v6780khuazyt))
- Define metrics if necessary, and provide helper functions for writing metrics from scrapers in the new `metrics` module
- Add metrics for the data types described in the design doc's [Monitoring section](https://docs.google.com/document/d/11fv-hLwLOFcM5XSE16lYne5hnqFzcQK1vwvTBVVyhn8/edit#heading=h.pq2010l3u22n)

**Comment by colincadams** [comment]
Just found this thread. Based on [this discussion in OpenCensus](https://github.com/census-instrumentation/opencensus-go/issues/576), we should be able to handle alerting on cumulative metrics by creating some gauge views of them. Should work for now and hopefully in the future it will be supported fully.

**Comment by colincadams** [comment]
stealing this issue for now to track monitoring

**Comment by colincadams** [comment]
A first step is already in with #449, but it actually functioning is blocked by #596

**Comment by arian487** [comment]
Whenever https://github.com/Recidiviz/recidiviz-data/pull/640 goes in, along with another PR handling error reporting on the ingest side I'm going to consider this done for the purposes of drop day as that is the most important.  We can add these other metrics over time.  Moving this out for now.

**Comment by jessex** [comment]
Where do we stand on this one now, @colincadams and @arian487?

**Comment by colincadams** [comment]
#1394 adds a metric for scraper work, then I'm planning to add one for batch persistence.

I think that will be a reasonable base set and will allow us to tell if and where failures are happening. Then plenty more metrics can be added throughout the code, especially if we are interested in tracking something specific (i.e. latency of a crucial/volatile step like entity matching), but I don't think they are essential at this point and we likely won't get to them.

I'll add links to the dashboards, some base alerts (if we are stable enough) and docs/playbooks and then plan to close this out mid next week. WDYT?

**Comment by jessex** [comment]
Sounds great. It would be lovely to also document offline what else you'd look to instrument given the time, if we don't already have that list. I have an old list myself but it may no longer be up-to-date.
